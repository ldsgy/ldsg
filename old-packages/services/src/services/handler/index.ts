import { Service, SettingsSchema } from "@ldsg/core";
import { HANDLER_SERVICE_DEFAULT_SETTINGS_SCHEMA } from "./constants";

export * from "./constants";

type Handler = (...rest: any[]) => any | Promise<any>;

type GetHandler = () => Handler;

export interface HandlerInfo {
  requireKey: string;
}

export class HandlerService extends Service {
  settingsSchema: SettingsSchema = HANDLER_SERVICE_DEFAULT_SETTINGS_SCHEMA;

  getHandler: GetHandler = () => {
    const moduleId = this.getModuleId();

    try {
      const requireResult = require(moduleId);

      if ("handler" in requireResult) {
        return requireResult.handler;
      } else {
        throw new Error("invalid handler");
      }
    } catch (error) {
      console.error(error);

      throw new Error("invalid require");
    }
  };

  getModuleId = (): string => {
    const settings = this.getSettings();

    const { moduleId } = settings;

    const res: string = moduleId;

    return res;
  };
}
