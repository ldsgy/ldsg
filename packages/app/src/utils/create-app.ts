import {
  ApplicationResource,
  resourceDefinitionResourceSettings as applicationResourceDefinitionResourceSettings,
} from "@ldsg/application";
import { Resource, ResourceRecord } from "@ldsg/resource";
import { instantiateResourcesRequiredByApp } from "@ldsg/resource-instantiator";
import { Express } from "express";

interface CreateAppParams {
  resources?: Resource[];
  resourceRecords?: ResourceRecord[];
}

type CreateApp = (params: CreateAppParams) => Express;

export const createApp: CreateApp = (params) => {
  const { resources: paramsResources, resourceRecords } = params;

  let resources: Resource[];

  if (paramsResources) {
    resources = paramsResources;
  } else {
    if (!resourceRecords) {
      throw new Error("invalid manifest resources");
    }

    const resolveManifestResourcesRes = instantiateResourcesRequiredByApp({
      resourceRecords,
    });

    resources = resolveManifestResourcesRes.resources;
  }

  const applicationResource = resources.find(
    (value) => value.kind === applicationResourceDefinitionResourceSettings.kind
  ) as ApplicationResource | undefined;

  if (!applicationResource) {
    throw new Error("invalid application resource");
  }

  const res = applicationResource.createApp();

  return res;
};
