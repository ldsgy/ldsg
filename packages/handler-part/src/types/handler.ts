export type DefaultParams = any[];

export type DefaultRes = any;

export type Handler<
  Params extends DefaultParams = DefaultParams,
  Res extends DefaultRes = DefaultRes
> = (...params: Params) => Res | Promise<Res>;
