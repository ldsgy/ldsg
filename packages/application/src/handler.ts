import { Handler } from "@ldsg/handler";
import { Manifest } from "@ldsg/resource";
import { ApplicationResource } from "./resource";
import { ApplicationResourceSettings } from "./types";

export const handler: Handler<
  [Manifest.Resource<ApplicationResourceSettings>]
> = (params) => {
  return new ApplicationResource(params);
};
