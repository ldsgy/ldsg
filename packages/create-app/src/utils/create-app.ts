import {
  RESOURCE_DEFINITION_SPECIFIC_RESOURCE_SETTINGS as APPLICATION_RESOURCE_DEFINITION_SPECIFIC_RESOURCE_SETTINGS,
  ApplicationResource,
} from "@ldsg/application";
import { Resource } from "@ldsg/resource";
import { instantiateResourcesRequiredByApp } from "@ldsg/resource-instantiator";
import { ResourceRecord } from "@ldsg/types";
import { Express } from "express";

interface CreateAppParams {
  resources?: Resource[];
  resourceRecords?: ResourceRecord[];
}

type CreateApp = (params: CreateAppParams) => Express;

export const createApp: CreateApp = (params) => {
  const { resources: paramsResources, resourceRecords } = params;

  let resources: Resource[];

  if (paramsResources) {
    resources = paramsResources;
  } else {
    if (!resourceRecords) {
      throw new Error("invalid manifest resources");
    }

    const resolveManifestResourcesRes = instantiateResourcesRequiredByApp({
      resourceRecords,
    });

    resources = resolveManifestResourcesRes.resources;
  }

  const applicationResource = resources.find(
    (value) =>
      value.kind ===
      APPLICATION_RESOURCE_DEFINITION_SPECIFIC_RESOURCE_SETTINGS.kind
  ) as ApplicationResource | undefined;

  if (!applicationResource) {
    throw new Error("invalid application resource");
  }

  const res = applicationResource.createExpressApp();

  return res;
};
