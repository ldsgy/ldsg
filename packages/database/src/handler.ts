import { Handler } from "@ldsg/handler";
import {
  InstantiateResourceParams,
  InstantiateResourceRes,
} from "@ldsg/resource";
import { instantiateResource } from "./instantiate-resource";
import { DatabaseResource } from "./resource";
import { DatabaseSpecificResourceSettings } from "./types";

export const handler: Handler<
  [InstantiateResourceParams<DatabaseSpecificResourceSettings>],
  InstantiateResourceRes<DatabaseResource>
> = instantiateResource;
