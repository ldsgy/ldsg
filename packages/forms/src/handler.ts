import { Handler, HandlerExtendedResourceSettings } from "@ldsg/handler";
import {
  InstantiateResourceParams,
  InstantiateResourceRes,
} from "@ldsg/resource";
import { instantiateResource } from "./instantiate-resource";
import { FormResource } from "./resource";
import { FormSpecificResourceSettings } from "./types";

export const handler: Handler<
  [
    InstantiateResourceParams<
      HandlerExtendedResourceSettings<FormSpecificResourceSettings>
    >
  ],
  InstantiateResourceRes<FormResource>
> = instantiateResource;
