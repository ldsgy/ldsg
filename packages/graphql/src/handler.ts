import { Handler, HandlerExtendedResourceSettings } from "@ldsg/handler";
import {
  InstantiateResourceParams,
  InstantiateResourceRes,
} from "@ldsg/resource";
import { instantiateResource } from "./instantiate-resource";
import { GraphqlResource } from "./resource";
import { GraphqlSpecificResourceSettings } from "./types";

export const handler: Handler<
  [
    InstantiateResourceParams<
      HandlerExtendedResourceSettings<GraphqlSpecificResourceSettings>
    >
  ],
  InstantiateResourceRes<GraphqlResource>
> = instantiateResource;
