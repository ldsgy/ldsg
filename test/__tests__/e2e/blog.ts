import { createApp } from "@ldsg/create-app";
import { RESOURCE_DEFINITION_SPECIFIC_RESOURCE_SETTINGS as HANDLER_RESOURCE_DEFINITION_SPECIFIC_RESOURCE_SETTINGS } from "@ldsg/handler";
import request from "supertest";
import { transpile } from "typescript";
import vm from "vm";
import { blogResourceRecords as resourceRecords } from "../../apps";

const runIsolated = (code: string): any => {
  // 创建新的沙箱对象，包含独立的 exports 和 module
  const sandbox = {
    exports: {},
    require: require,
    module: { exports: {} },
    console: console,
  };
  // 确保 module.exports 引用正确
  sandbox.module.exports = sandbox.exports;

  // 创建上下文并运行代码
  vm.createContext(sandbox);
  vm.runInContext(code, sandbox);

  // 返回当前沙箱的导出结果
  return sandbox.exports;
};

describe("blog", () => {
  test("resource records is completed", () => {
    expect(resourceRecords).toMatchSnapshot();
  });

  const resourceRecordsWithHandlerFunction = resourceRecords.map(
    (resourceRecord) => {
      const { kind, settings } = resourceRecord;

      let res = resourceRecord;

      switch (kind) {
        case HANDLER_RESOURCE_DEFINITION_SPECIFIC_RESOURCE_SETTINGS.kind: {
          const { code: tsCode, handler } = settings;

          if (!handler) {
            const jsCode = transpile(tsCode);

            const handler = runIsolated(jsCode).handler;

            res = {
              ...resourceRecord,
              settings: {
                ...settings,
                handler,
              },
            };
          }

          break;
        }

        default: {
          break;
        }
      }

      return res;
    }
  );

  describe("app", () => {
    const app = createApp({
      resourceRecords: resourceRecordsWithHandlerFunction,
    });

    test("is health", async () => {
      const response = await request(app).get("/health");

      expect(response.statusCode).toBe(200);

      expect(response.text).toBe("ok");
    });

    test("graphql is health", async () => {
      const response = await request(app).get("/graphql");

      expect(response.statusCode).toBe(200);

      expect(response.text).toMatchSnapshot();
    });
  });
});
