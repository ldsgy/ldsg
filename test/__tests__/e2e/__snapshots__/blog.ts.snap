// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blog app graphql is health 1`] = `"{"errors":[{"message":"Must provide query string.","extensions":{"code":"BAD_REQUEST"}}]}"`;

exports[`blog resource records is completed 1`] = `
[
  {
    "id": "application-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "",
      "handlerResourceId": "application-resource-definition-handler",
      "kind": "application",
      "parentKind": "root",
      "title": "应用类型资源定义",
    },
  },
  {
    "id": "application-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/application";",
      "dependencies": [
        {
          "name": "@ldsg/application",
        },
      ],
      "description": "",
      "title": "应用类型资源处理程序",
    },
  },
  {
    "id": "database-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "",
      "handlerResourceId": "database-resource-definition-handler",
      "kind": "database",
      "parentKind": "root",
      "title": "数据库类型资源定义",
    },
  },
  {
    "id": "database-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/database";",
      "dependencies": [
        {
          "name": "@ldsg/database",
        },
      ],
      "description": "",
      "title": "数据库类型资源处理程序",
    },
  },
  {
    "id": "field-type-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "",
      "handlerResourceId": "field-type-resource-definition-handler",
      "kind": "field_type",
      "parentKind": "root",
      "title": "字段类型类型资源定义",
    },
  },
  {
    "id": "field-type-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/field-type";",
      "dependencies": [
        {
          "name": "@ldsg/field-type",
        },
      ],
      "description": "",
      "title": "字段类型类型资源处理程序",
    },
  },
  {
    "id": "form-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "",
      "handlerResourceId": "form-resource-definition-handler",
      "kind": "form",
      "parentKind": "forms",
      "title": "表单资源定义",
    },
  },
  {
    "id": "form-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/form";",
      "dependencies": [
        {
          "name": "@ldsg/form",
        },
      ],
      "description": "",
      "title": "表单资源处理程序",
    },
  },
  {
    "id": "forms-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "",
      "handlerResourceId": "forms-resource-definition-handler",
      "kind": "forms",
      "parentKind": "application",
      "title": "表单组类型资源定义",
    },
  },
  {
    "id": "forms-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/forms";",
      "dependencies": [
        {
          "name": "@ldsg/forms",
        },
      ],
      "description": "",
      "title": "表单组类型资源处理程序",
    },
  },
  {
    "id": "graphql-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "",
      "handlerResourceId": "graphql-resource-definition-handler",
      "kind": "graphql",
      "parentKind": "application",
      "title": "GraphQL类型资源定义",
    },
  },
  {
    "id": "graphql-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/graphql";",
      "dependencies": [
        {
          "name": "@ldsg/graphql",
        },
      ],
      "description": "",
      "title": "GraphQL类型资源处理程序",
    },
  },
  {
    "id": "handler-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "主要包含引入模块列表与相应的处理程序代码，处理程序类型资源无子级资源。",
      "handlerResourceId": "handler-resource-definition-handler",
      "kind": "handler",
      "parentKind": "root",
      "title": "处理程序类型资源定义",
    },
  },
  {
    "id": "handler-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/handler";",
      "dependencies": [
        {
          "name": "@ldsg/handler",
        },
      ],
      "description": "包含处理程序相应的模块，此模块主要通过",
      "title": "处理程序类型资源处理程序",
    },
  },
  {
    "id": "object-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "",
      "handlerResourceId": "object-resource-definition-handler",
      "kind": "object",
      "parentKind": "root",
      "title": "对象类型资源定义",
    },
  },
  {
    "id": "object-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/object";",
      "dependencies": [
        {
          "name": "@ldsg/object",
        },
      ],
      "description": "",
      "title": "对象类型资源处理程序",
    },
  },
  {
    "id": "object-field-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "",
      "handlerResourceId": "object-field-resource-definition-handler",
      "kind": "object_field",
      "parentKind": "object",
      "title": "对象字段类型资源定义",
    },
  },
  {
    "id": "object-field-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/object-field";",
      "dependencies": [
        {
          "name": "@ldsg/object-field",
        },
      ],
      "description": "",
      "title": "对象字段类型资源处理程序",
    },
  },
  {
    "id": "resource-definition-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "",
      "handlerResourceId": "resource-definition-resource-definition-handler",
      "kind": "resource_definition",
      "parentKind": "root",
      "title": "资源定义类型资源定义",
    },
  },
  {
    "id": "resource-definition-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/resource-definition";",
      "dependencies": [
        {
          "name": "@ldsg/resource-definition",
        },
      ],
      "description": "",
      "title": "资源定义类型资源处理程序",
    },
  },
  {
    "id": "route-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "",
      "handlerResourceId": "route-resource-definition-handler",
      "kind": "route",
      "parentKind": "application",
      "title": "路由类型资源定义",
    },
  },
  {
    "id": "route-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/route";",
      "dependencies": [
        {
          "name": "@ldsg/route",
        },
      ],
      "description": "",
      "title": "路由类型资源处理程序",
    },
  },
  {
    "id": "table-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "",
      "handlerResourceId": "table-resource-definition-handler",
      "kind": "table",
      "parentKind": "tables",
      "title": "表格资源定义",
    },
  },
  {
    "id": "table-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/table";",
      "dependencies": [
        {
          "name": "@ldsg/table",
        },
      ],
      "description": "",
      "title": "表格资源处理程序",
    },
  },
  {
    "id": "tables-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "",
      "handlerResourceId": "tables-resource-definition-handler",
      "kind": "tables",
      "parentKind": "application",
      "title": "表格组类型资源定义",
    },
  },
  {
    "id": "tables-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/tables";",
      "dependencies": [
        {
          "name": "@ldsg/tables",
        },
      ],
      "description": "",
      "title": "表格组类型资源处理程序",
    },
  },
  {
    "id": "workflow-node-type-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "",
      "handlerResourceId": "workflow-node-type-resource-definition-handler",
      "kind": "workflow_node_type",
      "parentKind": "root",
      "title": "工作流类型资源定义",
    },
  },
  {
    "id": "workflow-node-type-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/workflow";",
      "dependencies": [
        {
          "name": "@ldsg/workflow",
        },
      ],
      "description": "",
      "title": "工作流类型资源处理程序",
    },
  },
  {
    "id": "workflow-node-type-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "",
      "handlerResourceId": "workflow-node-type-resource-definition-handler",
      "kind": "workflow_node_type",
      "parentKind": "workflow_node_type",
      "title": "工作流节点类型类型资源定义",
    },
  },
  {
    "id": "workflow-node-type-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/workflow-edge";",
      "dependencies": [
        {
          "name": "@ldsg/workflow-edge",
        },
      ],
      "description": "",
      "title": "工作流节点类型类型资源处理程序",
    },
  },
  {
    "id": "workflow-node-type-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "",
      "handlerResourceId": "workflow-node-type-resource-definition-handler",
      "kind": "workflow_node_type",
      "parentKind": "workflow_node_type",
      "title": "工作流节点类型资源定义",
    },
  },
  {
    "id": "workflow-node-type-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/workflow-node";",
      "dependencies": [
        {
          "name": "@ldsg/workflow-node",
        },
      ],
      "description": "",
      "title": "工作流节点类型资源处理程序",
    },
  },
  {
    "id": "workflow-node-type-resource-definition",
    "kind": "resource_definition",
    "parentId": "root",
    "settings": {
      "description": "",
      "handlerResourceId": "workflow-node-type-resource-definition-handler",
      "kind": "workflow_node_type",
      "parentKind": "root",
      "title": "工作流节点类型类型资源定义",
    },
  },
  {
    "id": "workflow-node-type-resource-definition-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "export * from "@ldsg/workflow-node-type";",
      "dependencies": [
        {
          "name": "@ldsg/workflow-node-type",
        },
      ],
      "description": "",
      "title": "工作流节点类型类型资源处理程序",
    },
  },
  {
    "id": "blog-application",
    "kind": "application",
    "parentId": "root",
    "settings": {
      "description": "",
      "title": "博客应用",
    },
  },
  {
    "id": "mian-graphql",
    "kind": "graphql",
    "parentId": "blog-application",
    "settings": {
      "description": "",
      "title": "GraphQL",
    },
  },
  {
    "id": "health-route-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "",
      "dependencies": [],
      "description": "",
      "handler": [Function],
      "title": "health 路由处理程序",
    },
  },
  {
    "id": "health-route",
    "kind": "route",
    "parentId": "blog-application",
    "settings": {
      "description": "",
      "handlerResourceId": "health-route-handler",
      "path": "/health",
      "title": "health 路由",
    },
  },
  {
    "id": "main-database",
    "kind": "database",
    "parentId": "root",
    "settings": {
      "description": "",
      "title": "主要数据库",
      "uri": "mongodb://root:example@mongo:27017/test?directConnection=true&authSource=admin",
    },
  },
  {
    "id": "array-field-type-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "",
      "dependencies": [],
      "description": "",
      "handler": [Function],
      "title": "数组字段类型处理程序",
    },
  },
  {
    "id": "array-field-type",
    "kind": "field_type",
    "parentId": "root",
    "settings": {
      "description": "",
      "fieldPropertiesSchema": {
        "properties": {},
        "type": "object",
      },
      "handlerResourceId": "array-field-type-handler",
      "title": "数组字段类型",
    },
  },
  {
    "id": "boolean-field-type-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "",
      "dependencies": [],
      "description": "",
      "handler": [Function],
      "title": "布尔字段类型处理程序",
    },
  },
  {
    "id": "boolean-field-type",
    "kind": "field_type",
    "parentId": "root",
    "settings": {
      "description": "",
      "fieldPropertiesSchema": {
        "properties": {},
        "type": "object",
      },
      "handlerResourceId": "boolean-field-type-handler",
      "title": "布尔字段类型",
    },
  },
  {
    "id": "mongo-id-field-type-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "",
      "dependencies": [],
      "description": "",
      "handler": [Function],
      "title": "MongoID字段类型处理程序",
    },
  },
  {
    "id": "mongo-id-field-type",
    "kind": "field_type",
    "parentId": "root",
    "settings": {
      "description": "",
      "fieldPropertiesSchema": {
        "properties": {},
        "type": "object",
      },
      "handlerResourceId": "mongo-id-field-type-handler",
      "title": "MongoID字段类型",
    },
  },
  {
    "id": "number-field-type-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "",
      "dependencies": [],
      "description": "",
      "handler": [Function],
      "title": "文本字段类型处理程序",
    },
  },
  {
    "id": "number-field-type",
    "kind": "field_type",
    "parentId": "root",
    "settings": {
      "description": "可用来存储各种文本",
      "fieldPropertiesSchema": {
        "properties": {
          "max": {
            "description": "",
            "title": "最大值",
            "type": "integer",
          },
          "min": {
            "description": "",
            "title": "最小值",
            "type": "integer",
          },
        },
        "type": "object",
      },
      "handlerResourceId": "number-field-type-handler",
      "title": "文本字段类型",
    },
  },
  {
    "id": "object-field-type-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "",
      "dependencies": [],
      "description": "",
      "handler": [Function],
      "title": "对象字段类型处理程序",
    },
  },
  {
    "id": "object-field-type",
    "kind": "field_type",
    "parentId": "root",
    "settings": {
      "description": "",
      "fieldPropertiesSchema": {
        "properties": {},
        "type": "object",
      },
      "handlerResourceId": "object-field-type-handler",
      "title": "对象字段类型",
    },
  },
  {
    "id": "string-field-type-handler",
    "kind": "handler",
    "parentId": "root",
    "settings": {
      "code": "",
      "dependencies": [],
      "description": "",
      "handler": [Function],
      "title": "字符串字段类型处理程序",
    },
  },
  {
    "id": "string-field-type",
    "kind": "field_type",
    "parentId": "root",
    "settings": {
      "description": "",
      "fieldPropertiesSchema": {
        "properties": {},
        "type": "object",
      },
      "handlerResourceId": "string-field-type-handler",
      "title": "字符串字段类型",
    },
  },
  {
    "id": "workflow_node_type_end",
    "kind": "workflow_node_type",
    "parentId": "root",
    "settings": {
      "description": "",
      "fieldPropertiesSchema": {},
      "handlerResourceId": "",
      "title": "结束",
    },
  },
  {
    "id": "workflow_node_type_start",
    "kind": "workflow_node_type",
    "parentId": "root",
    "settings": {
      "description": "",
      "fieldPropertiesSchema": {},
      "handlerResourceId": "",
      "title": "开始",
    },
  },
]
`;
